pkg/types/chain_reader.go

```diff
@@ -0,0 +1,74 @@
+package types
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type ChainReaderError string
+
+func (e ChainReaderError) Error() string { return string(e) }
+
+const (
+	ErrInvalidType   = ChainReaderError("invalid type")
+	ErrInvalidConfig = ChainReaderError("invalid configuration")
+)
+
+func UnwrapClientError(err error) error {
+	if s, ok := status.FromError(err); ok {
+		if s.Code() == codes.Unknown { // Only unwrap custom errors we return, leave alone any other gRPC generated error codes
```

Comment:
🤔 "invalid type" and "invalid configuration" feels like they would correspond to GRPC's INVALID_ARGUMENT status code, they aren't really unknown.

In general, intercepting codes.Unknown and interpreting that as an application level error feels a little wrong. I would recommend mapping your application errors to one of the standard error codes, and if necessary pass in additional metadata via the errdetails package: https://pkg.go.dev/google.golang.org/genproto/googleapis/rpc/errdetails

Reply:
@reductionista  I think you went over this with Jordan

Reply:
I was looking at the relationship between `errors.Unsupported` and `codes.Unimplemented`, but I'm curious about this too.

Reply:
At least for invalid type, I was trying to contain the errors that chain reader (and later the codec) will have.  They don't know if they are being run in a LOOP.

Reply:
My proposal:

If it's okay to use errors formatted using `status.Errorf` whether or not we're running in gRPC then I think we should just get rid of `UnwrapClientError` and use `codes.InvalidArgument` for both `InvalidType` and `InvalidConfig`.  We may as well also get rid of the `ChainlinkReaderError` itself and just have these errors defined as strings.

If we should only use `status.Errorf` in the gRPC layer itself (client & server methods, rather than the implementation) then we do need to preserve some sort of wrap/unwrap helpers.  But the logic needs to be more complex, so that it intercepts specific error types and maps them to gRPC codes and back again.

Reply:
Ideally we would only wrap and unwrap the grpc errors inside the grpc clients and servers. If we depend on detecting certain kinds of errors on the client side, but from outside the client, then we need to find a way to preserve those checks, regardless of whether it goes over grpc or not. Have we considered `FromError()`?
https://pkg.go.dev/google.golang.org/grpc/status#FromError
> FromError returns a Status representation of err.
>
> - If err was produced by this package or implements the method `GRPCStatus() *Status` and `GRPCStatus()` does not return nil, or if err wraps a type satisfying this, the Status from `GRPCStatus()` is returned. For wrapped errors, the message returned contains the entire err.Error() text and not just the wrapped status. In that case, ok is true.

It sounds like we could have our server side custom error types implement this method to support sending themselves over the wire with the details that Cedric linked. I'm not sure how to correctly 'unwrap' on the other end though.

Reply:
Yeah... that seems like the best approach for the wrap side.  Although it still doesn't allow for reconstructing wrapped errors easily, we'd have to implement our own system of parsing if we want to handle that.

If we're doing it this way, I think it might make sense after all to go back to converting `errors.Unsupported` into `code.Unimplemented` and back.  We'd just have to implement `func ( err errors.ErrUnsupported) GRPCStatus() *Status`.

On the unwrap side, the thing which would make the most sense to me would be to have one error type per gRPC code (at least, the ones we use) and convert it into that error type with a value extracted from the status object in whatever way is appropriate.

One thing I noticed while looking through the code today is that the `Unimplemented` code is supposed to be fairly broad, covering both unsupported methods and disabled methods:
```
// The operation is not implemented or is not supported/enabled in this
	// service.
	//
	// HTTP Mapping: 501 Not Implemented
	[Code](https://pkg.go.dev/google.golang.org/genproto/googleapis/rpc@v0.0.0-20231120223509-83a465c0220f/code#Code)_UNIMPLEMENTED Code = 12
```
https://pkg.go.dev/google.golang.org/genproto/googleapis/rpc@v0.0.0-20231120223509-83a465c0220f/code

This doesn't matter all that much for now, because presently when `ChainReader` is disabled we just fall back.  But after we rip out the old code, it will be a requirement in order to create a feeds job.  Getting this error code back could either mean it's disabled in the job spec or that it's an older version of the relay where it's not implemented at all.  The details of the status object would contain the specific reason... either just in the `Message` field or in a more complex way as part of the payload.  I could also see a case for making the disabled case a separate error type, and having that one correspond to a `FAILED_PRECONDITION`.  Either way the message attached to it would say "ChainReader disabled in RelayConfig".

Reply:
We have the option of extending our custom error type with an Is method that gets called by errors.Is. this lets us define our own version of equality, so we could have it return true for GRPC errors with the appropriate code and details. This would avoid the need for a magic unwrap function and maybe simplify things overall.

Reply:
I committed an implementation of the above to `chainlink-common`.  It wasn't that big of a change, instead of `ChainReaderError` type we have error types `InvalidArgumentError` and `UnimplementedError`.  I moved them both to plugin_provider.go because it seems like we could use this for any provider.  Specific types are defined in the same way as before, but now the status code can get set based on the error type and the original error type chosen based on the status code received on the client side.

At first instead of defining our own `UnimplementedError` I used `errors.ErrUnsupported`.  This works fine if it's just `errors.ErrUnsupported` by itself, but the error message associated with that is simply "operation not supported".  If you try to add any more detail, by wrapping another error with it then it behaves badly. The error gets modified by the gRPC layer as it's passing through, starting out on the server side as something which passes `errors.Is(error.Unsupported)` but failing to satisfy that on the client side. I think that's something we want to avoid in general, because it means the way the error behaves will be different depending on whether LOOP is enabled or not.  The simple fix to that was to just to define it as another `string` type along side all the other error types we define (which presently is only `InvalidArgumentError`, but we can add the rest as we need them).